core/node/node_storage_init/src/traits.rs

```diff
@@ -0,0 +1,30 @@
+use std::fmt;
+
+use tokio::sync::watch;
+use zksync_types::L1BatchNumber;
+
+/// An abstract storage initialization strategy.
+#[async_trait::async_trait]
+pub trait InitializeStorage: fmt::Debug + Send + Sync + 'static {
+    /// Checks if the storage is already initialized.
+    async fn is_initialized(&self) -> anyhow::Result<bool>;
+
+    /// Initializes the storage.
+    /// Implementors of this method may assume that they have unique access to the storage.
+    async fn initialize_storage(&self, stop_receiver: watch::Receiver<bool>) -> anyhow::Result<()>;
+}
+
+/// An abstract storage revert strategy.
+/// This trait assumes that for any invalid state there exists a batch number to which the storage can be rolled back.
+#[async_trait::async_trait]
+pub trait RevertStorage: fmt::Debug + Send + Sync + 'static {
+    /// Checks if the storage is invalid state and has to be rolled back.
+    async fn is_revert_needed(&self) -> anyhow::Result<Option<L1BatchNumber>>;
```

Comment:
Bikeshedding: Since this method returns richer data than `bool`, IMO, it should be named like `last_correct_l1_batch_for_reorg`, or `detect_reorg` + use a dedicated struct wrapping `L1BatchNumber`:

```rust
pub struct DetectedReorg {
    pub last_correct_l1_batch: L1BatchNumber,
}
```

Reply:
Changed the name but kept the type; IMHO it's self-explanatory enough, and if not, there exists a doc comment for that.